### PR TITLE
DROID-2793 Date as an Object | Epic

### PR DESCRIPTION
**Pull request by konstantiniiv**
---

- [x] I understand that contributing to this repository will require me to agree with the [CLA](https://github.com/anyproto/.github/blob/main/docs/CLA.md)

---

### What type of PR is this? (check all applicable)

- [x] 🍕 Feature
- [ ] 🐛 Bug Fix
- [ ] 📝 Documentation Update
- [ ] 🎨 Style
- [ ] 🧑‍💻 Code Refactor
- [ ] 🔥 Performance Improvements
- [ ] ✅ Test
- [ ] 🤖 Build
- [ ] 🔁 CI

### Added tests?

- [ ] 👍 yes
- [ ] 🙅 no, because they aren't needed
- [ ] 🙋 no, because I need help

### Added to documentation?

- [ ] 📜 README.md
- [ ] 📓 [tech-docs](https://github.com/anyproto/tech-docs)
- [ ] 🙅 no documentation needed

### CODE REVIEW

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+                                space = space
+                            )
+                        )
+                    }.onFailure { e ->
+                        Timber.e(e, "Error while opening global search screen from all content")
+                    }
+                }
+                is DateEffect.SendToast.Error -> TODO()
+                is DateEffect.SendToast.ObjectArchived -> TODO()
+                is DateEffect.NavigateToDateObject -> {
```
Comment (uburoiubu) [inline app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:132]: might be a legacy

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+                    buttonText = stringResource(id = R.string.membership_error_button_text_dismiss),
+                    onButtonClick = vm::hideError,
+                    onDismissRequest = vm::hideError
+                )
+            }
+        }
+    }
+
+    override fun injectDependencies() {
+        val params = DateVmParams(
```
Comment (uburoiubu) [inline app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:240]: `DateObjectVmParams`

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+        )
+        componentManager().dateObjectComponent.get(params).inject(this)
+    }
+
+    override fun releaseDependencies() {
+        componentManager().dateObjectComponent.release()
+    }
+
+    override fun onApplyWindowRootInsets(view: View) {
+        // Do nothing. TODO add ime padding.
```
Comment (uburoiubu) [inline app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:252]: may need to check on older android versions

File: app/src/main/java/com/anytypeio/anytype/ui/vault/VaultFragment.kt [hunk 4]
```diff
@@ -168,6 +167,13 @@ class VaultFragment : BaseComposeFragment() {
                     space = destination.space
                 )
             }
+            is Navigation.OpenDateObject -> {
+                findNavController().navigate(R.id.actionOpenSpaceFromVault)
```
Comment (uburoiubu) [inline app/src/main/java/com/anytypeio/anytype/ui/vault/VaultFragment.kt:171]: `runCatching { ... }`

File: device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt [hunk 5]
```diff
@@ -101,16 +125,24 @@ class DateProviderImplTest {
             Triple(timeStamp, ZoneId.of("GMT+12"), 1720785600L),
             Triple(timeStamp, ZoneId.of("GMT-12"), 1720872000L)
         )
-        tests.forEach { (utcTimestamp, zoneId, expected) ->
-            val startOfDayInLocalZone =
-                dateProviderImpl.adjustFromStartOfDayInUserTimeZoneToUTC(utcTimestamp, zoneId)
-
-            assertEquals(expected, startOfDayInLocalZone)
-        }
+        // TODO fix tests
```
Comment (uburoiubu) [inline device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt:128]: iidem

File: domain/src/main/java/com/anytypeio/anytype/domain/objects/ObjectDateByTimestamp.kt [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+import com.anytypeio.anytype.core_models.Command
+import com.anytypeio.anytype.core_models.Struct
+import com.anytypeio.anytype.core_models.primitives.SpaceId
+import com.anytypeio.anytype.domain.base.AppCoroutineDispatchers
+import com.anytypeio.anytype.domain.base.ResultInteractor
+import com.anytypeio.anytype.domain.block.repo.BlockRepository
+import com.anytypeio.anytype.domain.objects.ObjectDateByTimestamp.Params
+import javax.inject.Inject
+
+class ObjectDateByTimestamp @Inject constructor(
```
Comment (uburoiubu) [inline domain/src/main/java/com/anytypeio/anytype/domain/objects/ObjectDateByTimestamp.kt:12]: `GetDateObjectByTimestamp`

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateEffect.kt [hunk 7]
```diff
@@ -0,0 +1,21 @@
+package com.anytypeio.anytype.feature_date.viewmodel
+
+import com.anytypeio.anytype.core_models.Id
+import com.anytypeio.anytype.core_models.primitives.SpaceId
+
+sealed class DateEffect {
```
Comment (uburoiubu) [inline feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateEffect.kt:6]: `DateObjectCommand`

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateModels.kt [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+sealed class UiObjectsListItem {
+
+    abstract val id: String
+
+    data class Loading(override val id: String) : UiObjectsListItem()
+
+    data class Item(
+        override val id: String,
+        val name: String,
+        val space: SpaceId?,
```
Comment (uburoiubu) [inline feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateModels.kt:133]: maybe turn space id into required field?

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,865 @@
...
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+/**
+ * ViewState: @see [UiContentState]
+ * Factory: @see [DateVMFactory]
+ * Screen: @see [com.anytypeio.anytype.feature_date.ui.DateMainScreen]
+ * Models: @see [UiObjectsListState]
+ */
+class DateViewModel(
```
Comment (uburoiubu) [inline feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt:73]: `DateObjectViewModel` or `DateAsObjectViewModel`

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/SearchParams.kt [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+        ),
+        DVFilter(
+            relation = Relations.IS_HIDDEN_DISCOVERY,
+            condition = DVFilterCondition.NOT_EQUAL,
+            value = true
+        )
+    )
+}
+
+private val SUPPORTED_LAYOUTS = listOf(
```
Comment (uburoiubu) [inline feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/SearchParams.kt:116]: `SUPPORTED_DATE_OBJECT_LAYOUTS`

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt [hunk 11]
```diff
@@ -0,0 +1,865 @@
...
+    @OptIn(ExperimentalCoroutinesApi::class)
+    private fun setupUiStateFlow() {
+        viewModelScope.launch {
+            combine(
+                _dateId.filterNotNull(),
+                _dateTimestamp.filterNotNull(),
+                _activeField.filterNotNull(),
+                restartSubscription
+            ) { dateId, timestamp, activeField, _ ->
+                loadData(dateId = dateId, timestamp = timestamp, field = activeField)
```
Comment (uburoiubu) [inline feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt:356]: we could map at this point `dateId, timestamp, activeField, _ ->` into search params. then load data when next operator is applied: `.flatMapLatest { searchParams -> loadData(searchParams) }`

**Review by uburoiubu — APPROVED** [review]

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt (hunk 1 above)
Comment (konstantiniiv) [reply app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:132]: done

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt (hunk 2 above)
Comment (konstantiniiv) [reply app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:240]: done

File: app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt (hunk 3 above)
Comment (konstantiniiv) [reply app/src/main/java/com/anytypeio/anytype/ui/date/DateObjectFragment.kt:252]: will do

File: app/src/main/java/com/anytypeio/anytype/ui/vault/VaultFragment.kt (hunk 4 above)
Comment (konstantiniiv) [reply app/src/main/java/com/anytypeio/anytype/ui/vault/VaultFragment.kt:171]: done

File: domain/src/main/java/com/anytypeio/anytype/domain/objects/ObjectDateByTimestamp.kt (hunk 6 above)
Comment (konstantiniiv) [reply domain/src/main/java/com/anytypeio/anytype/domain/objects/ObjectDateByTimestamp.kt:12]: done

File: domain/src/main/java/com/anytypeio/anytype/domain/relations/RelationListWithValue.kt [hunk 12]
```diff
@@ -0,0 +1,26 @@
...
+import com.anytypeio.anytype.core_models.Command
+import com.anytypeio.anytype.core_models.RelationListWithValueItem
+import com.anytypeio.anytype.core_models.primitives.SpaceId
+import com.anytypeio.anytype.domain.base.AppCoroutineDispatchers
+import com.anytypeio.anytype.domain.base.ResultInteractor
+import com.anytypeio.anytype.domain.block.repo.BlockRepository
+import com.anytypeio.anytype.domain.relations.RelationListWithValue.Params
+import javax.inject.Inject
+
+class RelationListWithValue @Inject constructor(
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply domain/src/main/java/com/anytypeio/anytype/domain/relations/RelationListWithValue.kt:12]: done

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateEffect.kt (hunk 7 above)
Comment (konstantiniiv) [reply feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateEffect.kt:6]: done

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateModels.kt (hunk 8 above)
Comment (konstantiniiv) [reply feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateModels.kt:133]: done

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt (hunk 9 above)
Comment (konstantiniiv) [reply feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt:73]: done

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/SearchParams.kt (hunk 10 above)
Comment (konstantiniiv) [reply feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/SearchParams.kt:116]: done

File: presentation/src/main/java/com/anytypeio/anytype/presentation/editor/EditorViewModel.kt [hunk 13]
```diff
@@ -6114,15 +6144,113 @@ class EditorViewModel(
...
+                spaceParams = provideParams(vmParams.space.id)
+            )
+            onCreateMentionInText(id = mention.id, name = mention.name, mentionTrigger = mentionTrigger)
+        }
+        if (mention is SelectDateItem) {
+            mentionDatePicker.value = EditorDatePickerState.Visible
+        }
+    }
+
+    private enum class PredefinedDate {
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply presentation/src/main/java/com/anytypeio/anytype/presentation/editor/EditorViewModel.kt:6163]: done

File: feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt (hunk 11 above)
Comment (konstantiniiv) [reply feature-date/src/main/java/com/anytypeio/anytype/feature_date/viewmodel/DateViewModel.kt]: fixed

File: device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt [hunk 14]
```diff
@@ -60,16 +76,24 @@ class DateProviderImplTest {
...
```
[the remark this answers is not in the document]
Comment (konstantiniiv) [reply device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt]: fixed

File: device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt (hunk 5 above)
Comment (konstantiniiv) [reply device/src/test/java/com/anytypeio/anytype/DateProviderImplTest.kt:120]: fixed

**Pull request merged**